pkg/volume/csi/csi_attacher.go

```diff
@@ -335,10 +496,80 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
 }
 
 func (c *csiAttacher) UnmountDevice(deviceMountPath string) error {
-	glog.V(4).Info(log("detacher.UnmountDevice is not implemented"))
+	glog.V(4).Info(log("attacher.UnmountDevice(%s)", deviceMountPath))
+
+	// Setup
+	ctx, cancel := grpctx.WithTimeout(grpctx.Background(), csiTimeout)
+	defer cancel()
+
+	driverName, volID, err := getDriverAndVolNameFromDeviceMountPath(deviceMountPath, c.plugin)
+	if err != nil {
+		glog.Errorf(log("attacher.UnmountDevice failed to get driver and volume name from device mount path: %v", err))
+		return err
+	}
+
+	if c.csiClient == nil {
+		addr := fmt.Sprintf(csiAddrTemplate, driverName)
+		c.csiClient = newCsiDriverClient("unix", addr)
+	}
+	csi := c.csiClient
+
+	// Check whether "STAGE_UNSTAGE_VOLUME" is set
+	capabilities, err := csi.NodeGetCapabilities(ctx)
+	if err != nil {
+		glog.Errorf(log("attacher.UnmountDevice failed while getting node capabilities: %v", err))
+		return err
+	}
+	if capabilities == nil {
+		glog.Infof(log("attacher.UnmountDevice STAGE_UNSTAGE_VOLUME capability not set. Skipping UnountDevice..."))
+		return nil
+	}
+
+	stageUnstageSet, err := hasStageUnstageCapability(capabilities)
+	if err != nil {
+		glog.Errorf(log("attacher.UnmountDevice failed: %v", err))
+		return err
+	}
+	if !stageUnstageSet {
+		glog.Infof(log("attacher.UnmountDevice STAGE_UNSTAGE_VOLUME capability not set. Skipping UnmountDevice..."))
+		return nil
+	}
+
+	// Start UnmountDevice
+	err = csi.NodeUnstageVolume(ctx,
+		volID,
+		deviceMountPath)
+
+	if err != nil {
+		glog.Errorf(log("attacher.UnmountDevice failed: %v", err))
+		return err
+	}
+
+	glog.V(4).Infof(log("attacher.UnmountDevice successfully requested NodeStageVolume [%s]", deviceMountPath))
 	return nil
 }
 
+func hasStageUnstageCapability(capabilities []*csipb.NodeServiceCapability) (bool, error) {
+	stageUnstageSet := false
+	for _, capability := range capabilities {
+		if capability.GetRpc().GetType() == csipb.NodeServiceCapability_RPC_STAGE_UNSTAGE_VOLUME {
+			stageUnstageSet = true
+		}
+	}
+	return stageUnstageSet, nil
+}
+
+func getDriverAndVolNameFromDeviceMountPath(deviceMountPath string, plugin *csiPlugin) (string, string, error) {
+	splitted := strings.Split(deviceMountPath, "/")
+	if len(splitted) < 4 {
+		return "", "", fmt.Errorf("DeviceMountPath malformed, got %s", deviceMountPath)
+	}
+	volID := strings.Join(splitted[len(splitted)-3:], "/")
+	dir := strings.Join(splitted[:len(splitted)-4], "/")
+	pluginName := plugin.host.GetPluginNameFromDir(dir)
```

Comment:
Assuming this`{pluginDir}/csi/{csiDriverName}/mounts/{csiVolumeHandle}` is the format, then  
`strings.Join(splitted[len(splitted)-3:], "/")` will return slice like `{csiDriverName}/mounts/{csiVolumeHandle}` .  Uneasy about the hardcoded offsets, seems a bit brittle.

Also, consider using filepath.Base(), filepath.Dir(), filepath.Join() for more portable path nomenclatures.

Reply:
I agree that the hardcoded offesets are brittle but I couldn't find a better way to do this without a hardcoded number of "Dir"'s anyway (3 of them). Because {csiVolumeHandle} itself is a 3 part object of form `project/zone/name`. That fact isn't really captured anywhere (maybe I should have a comment). Open to suggestions for better ways to do this.

I tried using `Rel` which ended up being more brittle because even small changes or differences to directory structure would completely destroy the resulting path.